### PR TITLE
Select kernel backends through BackendSpecProvider, not HAVE_* flags

### PR DESCRIPTION
**Pull request by guihong-nv**
## What

Every spec builder now asks a provider which implementation to use, instead of carrying its own availability flags and import guards.

**Stacked on #6762 — review that one first.** This PR's own change is 19 files, +449 / **−704**.

## The number this is for

Across the thirteen files that build specs:

| | before | after |
|---|---|---|
| `HAVE_TE` / `HAVE_APEX` / `HAVE_FUSED_LAYER_NORM` / `LNImpl` references | **72** | **0** |
| direct `transformer_engine` imports | **9** | **0** |

The files: `gpt_layer_specs`, `moe_module_specs`, `experimental_attention_variant_module_specs`, `multi_token_prediction`, `mla_qk_norm_config`, `language_module`, `models/backends`, `t5_spec`, `bert_layer_specs`, `bert_lm_head`, `vit_layer_specs`, `llava_spec`, `heterogeneous_layer_specs`.

## What is deliberately left alone

There are two populations of `HAVE_TE`, and only one of them is backend selection:

- **Selection (13 files) — fixed here.** `HAVE_TE` used to pick which module goes into a spec.
- **Implementation guards (28 files) — untouched.** `HAVE_TE` guarding `try: from transformer_engine.pytorch.X import Y`, where `Y` differs per file: `Float8Tensor`, `TransformerEngineBaseModule`, `get_dummy_wgrad`, `TE_DType`. Those files *wrap* TE rather than choosing between backends, each needs its own symbols bound, and they belong to the milestone that moves the TE wrappers.

Conflating the two is what makes the problem look like 112 files instead of 13.

## Selection is a plain if-else

```python
def get_backend(transformer_impl, *, use_kitchen=False, ...):
    base = _base_backend(transformer_impl, **settings)
    if not use_kitchen:
        return base
    return KitchenSpecProvider(fallback=base, ...)
```

and each provider is a flat list naming a class in `megatron.core.ops`, so reading `TESpecProvider` is the whole answer to which implementation a TE run gets.

## Deliberate behaviour changes

- **The `local` norm is Torch, not Apex-if-installed.** Installing an optional package no longer silently changes which kernel a run gets; Apex is chosen explicitly. This moves golden values for the 2 BERT functional cases and the GPT cases using `--spec local` — T5 has none.
- **`get_gpt_decoder_block_spec` passes `normalization` through** to the final norm, so `--transformer-impl local --normalization RMSNorm` stops asserting inside Apex's `FusedLayerNorm`.
- **Vocab-parallel cross entropy is chosen by the provider** rather than by a branch in the forward path, dropping an `UnboundLocalError` on one route and a `None` tensor-parallel group on another.

## Testing

8×GPU, all green: `ops/` 20, `test_bert_model` 14, `test_t5_model` 13, `test_llava_model` 24, `test_heterogeneous_gpt_model` 20, `test_clip_vit_model` 8, `test_gpt_model` 16, `test_moe_module_specs` 4, `test_transformer_block` 53, `test_cross_entropy` 3. Plus an import check of every rewritten spec module.

Needs a **Run functional tests** round trip for the golden values noted above.

### CODE REVIEW

**Comment by guihong-nv** [comment]
Auto-closed when #6762 closed (its base branch `pull-request/6762` was deleted), and GitHub will not reopen a PR whose base branch is gone. Continued in **#6766**, which is standalone and targets `main` directly.
